sklearn/semi_supervised/tests/test_self_training.py

```diff
@@ -0,0 +1,165 @@
+y_train_missing_labels[limit:] = -1
+y_train_missing_labels_strings = y_train_missing_labels.copy()
+mapping = {0: 'A', 1: 'B', 2: 'C', -1: -1}
+y_train_missing_strings = np.vectorize(mapping.get)(y_train_missing_labels)
+
+lb = LabelBinarizer()
+y_train_missing_dummies = lb.fit_transform(y_train_missing_labels)
+
+
+def test_classification():
+    # Check classification for various parameter settings.
+    # Also assert that predictions for strings and numerical labels are equal.
+    # Also test for multioutput classification
+    grid = ParameterGrid({"max_iter": [1, 50, 100],
+                          "threshold": [0.0, 0.5, 0.9]})
+
+    for base_classifier in [DummyClassifier(random_state=0),
+                            DecisionTreeClassifier(random_state=0),
+                            KNeighborsClassifier(),
+                            SVC(gamma="scale", probability=True,
+                                random_state=0)]:
+        for params in grid:
+            st = SelfTrainingClassifier(base_classifier, **params)
+            st_string = SelfTrainingClassifier(base_classifier, **params)
+            st.fit(X_train, y_train_missing_labels)
+            pred = st.predict(X_test)
+            proba = st.predict_proba(X_test)
+
+            st_string.fit(X_train, y_train_missing_strings)
+            pred_string = st_string.predict(X_test)
+            proba_string = st_string.predict_proba(X_test)
+
+            assert_array_equal(np.vectorize(mapping.get)(pred), pred_string)
+            assert_array_equal(proba, proba_string)
+
+
+def test_missing_predict_proba():
+    # Check that an error is thrown if predict_proba is not implemented
+    base_classifier = SVC(gamma="scale")
+    self_training = SelfTrainingClassifier(base_classifier)
+    message = "base_classifier (SVC) should implement predict_proba!"
+    assert_raise_message(ValueError, message, self_training.fit, X_train,
+                         y_train)
+
+
+def test_invalid_params():
+    # Test negative iterations
+    grid = ParameterGrid({"max_iter": [-1, -100, -10]})
+    base_classifier = SVC(gamma="scale", probability=True)
+    for params in grid:
+        st = SelfTrainingClassifier(base_classifier, **params)
+        message = "max_iter must be >= 0"
+        assert_raise_message(ValueError, message, st.fit, X_train, y_train)
+
+    grid = ParameterGrid({"threshold": [1.0, -2, 10]})
+    base_classifier = SVC(gamma="scale", probability=True)
+    for params in grid:
+        st = SelfTrainingClassifier(base_classifier, **params)
+        message = "threshold must be in [0,1)"
+        assert_raise_message(ValueError, message, st.fit, X_train, y_train)
+
+
+def test_single_iteration():
+    # Check classification for single iteration.
+    # Fitting a SelfTrainingClassifier with one iteration and 100 unlabeled
+    # datapoints should give the same results as fitting a normal classifier
+    # with only 50 labeled datapoints.
```

Comment:
You're not running one iteration here, you're running 0.

Also, the docstring is implicitely using fact that iris has 150 samples.

I think what you're trying to test here is that when you don't do any iteration, training `SelfTrainingClassifier` is the same as training the base classifier. The test is good but I would suggest making the docstring and the test name more explicit.

Reply:
Do you mean that I should remove mentions of the numbers from the docstrng?

Reply:
That, plus saying 0 instead of 1.